### ISSUE TITLE
bug(optimizer): unexpected array unnesting

### ISSUE THREAD

**Issue by KeXiangWang** [risingwavelabs/risingwave#14735]
### Error message/log

```text
?column? 
----------
 {2,3}
 {NULL}
```

### To Reproduce

```
create table t(a BIGINT[]);
insert into t values (ARRAY[3,2]::BIGINT[]), ('{}'::BIGINT[]);

SELECT array((SELECT a1 
FROM unnest(t.a) AS a1 ORDER BY a1)) from t;
``` 

### Expected behavior

```
 array
-------
 {2,3}
 {}
```

**Comment by xiangjinwu** [comment]
There are actually 2 issues here:

First, we rewrite `array(subquery)` into a scalar subquery with `array_agg` inside ([src](https://github.com/risingwavelabs/risingwave/blob/v1.7.0/src/frontend/src/optimizer/mod.rs#L165)):
```
SELECT array(SELECT a1 FROM unnest(t.a) AS a1 ORDER BY a1) from t;       -- old
SELECT (SELECT array_agg(a1 ORDER BY a1) FROM unnest(t.a) AS a1) from t; -- new
```
This step is not equivalent, as shown by PostgreSQL:
```
test=# SELECT array(SELECT a1 FROM unnest(t.a) AS a1 ORDER BY a1) from t;
 array 
-------
 {2,3}
 {}
(2 rows)

test=# SELECT (SELECT array_agg(a1 ORDER BY a1) FROM unnest(t.a) AS a1) from t;
 array_agg 
-----------
 {2,3}
 
(2 rows)
```

Second, `array_agg` is one of the few null-input-aware aggregates (`non-null input` vs `including nulls` in PostgreSQL [doc](https://www.postgresql.org/docs/current/functions-aggregate.html)) along with `count(*)` and `jsonb_agg`, and needs special treatment when transposing `Apply + SimpleAgg` into `GroupAgg + ApplyLeftOuter`. However, the current implementation ([src](https://github.com/risingwavelabs/risingwave/blob/v1.7.0/src/frontend/src/optimizer/rule/apply_agg_transpose_rule.rs#L67)) only handles `count(*)`.

Note that PostgreSQL has many variants of jsonb aggregates, as a result of duplication between pg-specific extensions and ISO SQL standard. At a glimpse, `jsonb_object_agg` / `jsonb_object_agg_strict` also needs such special treatment but it is unnecessary for `jsonb_agg_strict`.

cc @chenzl25

**Comment by xiangjinwu** [comment]
First issue of `array` rewrite fixed by #15593
Second issue of `array_agg` decorrelation fixed by #15590

Both `main` and `release-1.7` returns the expected output now.
